man/ettercap.8.in
.\"  ettercap -- a multipurpose sniffer/interceptor utility
.\"
.\"  This program is free software; you can redistribute it and/or modify
.\"  it under the terms of the GNU General Public License as published by
.\"  the Free Software Foundation; either version 2 of the License, or
.\"  (at your option) any later version.
.\"
.\"  This program is distributed in the hope that it will be useful,
.\"  but WITHOUT ANY WARRANTY; without even the implied warranty of
.\"  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\"  GNU General Public License for more details.
.\"
.\"  You should have received a copy of the GNU General Public License
.\"  along with this program; if not, write to the Free Software
.\"  Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
.\"
.de Sp
.if n .sp
.if t .sp 0.4
..
.TH ETTERCAP "8" "" "ettercap @VERSION@"
.SH NAME
ettercap - multipurpose sniffer/content filter for man in the middle attacks

.SH ***** IMPORTANT NOTE ******
Since ettercap NG (formerly 0.7.0), all the options have been changed. Even the target
specification has been changed. \fIPlease read carefully this man page.\fR

.SH SYNOPSIS
.B ettercap
[\fIOPTIONS\fR] [\fITARGET1\fR] [\fITARGET2\fR]
.PP
.B If IPv6 is enabled:
.br
.br
\fITARGET\fR is in the form \fIMAC/IPs/IPv6/PORTs\fR
.br
.B Otherwise,
.br
.br
\fITARGET\fR is in the form \fIMAC/IPs/PORTs\fR
.br
.br
where IPs and PORTs can be ranges (e.g. /192.168.0.1\-30,40,50/20,22,25)
.SH DESCRIPTION
Ettercap was born as a sniffer for switched LAN (and obviously even "hubbed" ones),
but during the development process it has gained more and more features that have
changed it to a powerful and flexible tool for man-in-the-middle attacks.
It supports active and passive dissection of many protocols (even ciphered ones)
and includes many features for network and host analysis (such as OS fingerprint).
.PP
It has two main sniffing options:
.PP
\fIUNIFIED\fR, this method sniffs all the packets that pass on the cable. You can choose to put or not the
interface in promisc mode (\-p option). The packet not directed to the host
running ettercap will be forwarded automatically using layer 3 routing. So you
can use a mitm attack launched from a different tool and let ettercap modify the
packets and forward them for you.
.br
The kernel ip_forwarding is always disabled by ettercap. This is done to
prevent a forward of a packet twice (one by ettercap and one by the kernel).
This is an invasive behaviour on gateways. So we recommend you to use ettercap
on the gateways ONLY with the UNOFFENSIVE MODE ENABLED. Since ettercap listens
only on one network interface, launching it on the gateway in offensive mode
will not allow packets to be rerouted back from the second interface.
.PP
\fIBRIDGED\fR, it uses two network interfaces and forward the traffic from one to the other
while performing sniffing and content filtering. This sniffing method is
totally stealthy since there is no way to find that someone is in the middle on
the cable.  You can look at this method as a mitm attack at layer 1. You will
be in the middle of the cable between two entities. Don't use it on gateways or
it will transform your gateway into a bridge. HINT: you can use the content
filtering engine to drop packets that should not pass. This way ettercap will
work as an inline IPS ;)
.PP
You can also perform man in the middle attacks while using the unified
sniffing. You can choose the mitm attack that you prefer. The mitm
attack module is independent from the sniffing and filtering process, so you can
launch several attacks at the same time or use your own tool for the attack. The
crucial point is that the packets have to arrive to ettercap with the correct
mac address and a different ip address (only these packets will be forwarded).
.PP
The most relevant ettercap features are:
.PP
.B SSH1 support :
you can sniff User and Pass, and even the data of an SSH1 connection. ettercap is the
first software capable to sniff an SSH connection in FULL-DUPLEX
.PP
.B SSL support :
you can sniff SSL secured data... a fake certificate is presented to the client and
the session is decrypted.
.PP
.B Characters injection in an established connection :
you can inject characters to the server (emulating commands) or to the client (emulating replies)
maintaining the connection alive !!
.PP
.B Packet filtering/dropping:
You can set up a filter script that searches for a particular string (even hex) in the TCP
or UDP payload and replace it with yours or drop the entire packet. The
filtering engine can match any field of the network protocols and modify
whatever you want (see etterfilter(8)).
.PP
.B Remote traffic sniffing through tunnels and route mangling:
You can play with linux cooked interfaces or use the integrated plugin to sniff
tunneled or route-mangled remote connections and perform mitm attacks on them.
.PP
.B Plug-ins support :
You can create your own plugin using the ettercap's API.
.PP
.B Password collector for :
TELNET, FTP, POP, RLOGIN, SSH1, ICQ, SMB, MySQL, HTTP, NNTP, X11, NAPSTER, IRC, RIP, BGP,
SOCKS 5, IMAP 4, VNC, LDAP, NFS, SNMP, HALF LIFE, QUAKE 3, MSN, YMSG (other protocols coming soon...)
.PP
.B Passive OS fingerprint:
you scan passively the lan (without sending any packet) and gather detailed info about
the hosts in the LAN: Operating System, running services, open ports, IP, mac address
and network adapter vendor.
.PP
.B Kill a connection:
from the connections list you can kill all the connections you want


.SH TARGET SPECIFICATION
There is no concept of SOURCE nor DEST. The purpose of the two targets is to
tell Ettercap which packets to focus on for traffic coming from one to the other
and vice-versa (since the connection is bidirectional).
.PP
\fITARGET\fR is in the form \fIMAC/IPs/PORTs\fR. 
.br
.B NOTE:
If IPv6 is enabled, \fITARGET\fR is in the form \fIMAC/IPs/IPv6/PORTs\fR.
.PP
If you want you can omit any of its parts and this will represent an ANY in that part.
.br
e.g.
.br
"//80" means ANY mac address, ANY ip and ONLY port 80
.br
"/10.0.0.1/" means ANY mac address, ONLY ip 10.0.0.1 and ANY port
.PP
\fIMAC\fR must be unique and in the form 00:11:22:33:44:55
.PP
\fIIPs\fR is a range of IP in dotted notation. You can specify range with the \- (hyphen)
and single ip with , (comma). You can also use ; (semicolon) to indicate
different ip addresses.
.br
e.g.
.br
"10.0.0.1\-5;10.0.1.33" expands into ip 10.0.0.1, 2, 3, 4, 5 and 10.0.1.33
.PP
\fIPORTs\fR is a range of PORTS. You can specify range with the \- (hyphen) and single port
with , (comma).
.br
e.g.
.br
"20\-25,80,110" expands into ports 20, 21, 22, 23, 24, 25, 80 and 110
.PP
.B NOTE:
.br
you can reverse the matching of the TARGET by adding the \-R option to the
command line. So if you want to sniff ALL the traffic BUT the one coming or
going to 10.0.0.1 you can specify "./ettercap \-R /10.0.0.1/"
.PP
.B NOTE:
.br
TARGETs are also responsible for the initial scan of the LAN. You can use them
to limit the scan to a subset of the hosts in the netmask.
The combined result of the two targets will be scanned.
Remember that not specifying a target means "no target", but specifying "//"
means "all the hosts in the subnet".
.PP


.SH PRIVILEGES DROPPING
ettercap needs root privileges to open the Link Layer sockets. After the
initialization phase, the root privs are not needed anymore, so ettercap drops
them to UID = 65535 (nobody). Since ettercap has to write (create) log files,
it must be executed in a directory with the right permissions (e.g. /tmp/). If
you want to drop privs to a different uid, you can export the environment
variable EC_UID with the value of the uid you want to drop the privs to (e.g.
export EC_UID=500) or set the correct parameter in the etter.conf file.


.SH SSL MITM ATTACK
SSL mitm attack is dependent on TCP traffic redirection to a custom listener
port of ettercap. The redir_command_on and redir_command_off configuration
variables take care of this (see "etter.conf(5)").
.br
However, when ettercap starts, traffic for any source and any destination
targeted for the redirectable services will be redirected to ettercap and the
SSL stream will be intercepted.
.br
This may not be the desired behaviour. Therefore you can adjust the redirect
rules after ettercap has been started using the selected user interface.
.Sp
While performing the SSL mitm attack, ettercap substitutes the real ssl
certificate with its own. The fake certificate is created on the fly and all
the fields are filled according to the real cert presented by the server. Only
the issuer is modified and signed with the private key contained in the 'etter.ssl.crt'
file. If you want to use a different private key you have to regenerate this
file. To regenerate the cert file use the following commands:
.Sp
openssl genrsa \-out etter.ssl.crt 1024
.br
openssl req \-new \-key etter.ssl.crt \-out tmp.csr
.br
openssl x509 \-req \-days 1825 \-in tmp.csr \-signkey etter.ssl.crt \-out tmp.new
.br
cat tmp.new >> etter.ssl.crt
.br
rm \-f tmp.new tmp.csr
.PP
NOTE: SSL mitm is not available (for now) in bridged mode.
.PP
NOTE: You can use the \-\-certificate/\-\-private\-key long options if you want to specify
a different file rather than the etter.ssl.crt file.


.SH OPTIONS
Options that make sense together can generally be combined. ettercap will warn the user
about unsupported option combinations.
.TP
.B SNIFFING AND ATTACK OPTIONS
.PP
ettercap NG has a new unified sniffing method. This implies that ip_forwarding
in the kernel is always disabled and the forwarding is done by ettercap. Every
packet with destination mac address equal to the host's mac address and
destination ip address different for the one bound to the iface will be
forwarded by ettercap. Before forwarding them, ettercap can content filter,
sniff, log or drop them. It does not matter how these packets are hijacked,
ettercap will process them. You can even use external programs to hijack
packet.
.br
You have full control of what ettercap should receive. You can use the internal
mitm attacks, set the interface in promisc mode, use plugins or use every
method you want.
.Sp
IMPORTANT NOTE: if you run ettercap on a gateway, remember to re-enable the
ip_forwarding after you have killed ettercap. Since ettercap drops its
privileges, it cannot restore the ip_forwarding for you.
.TP
\fB\-M\fR, \fB\-\-mitm <METHOD:ARGS>\fR
MITM attack
.br
This option will activate the man in the middle attack. The mitm attack is
totally independent from the sniffing. The aim of the attack is to hijack
packets and redirect them to ettercap. The sniffing engine will forward them if
necessary.
.br
You can choose the mitm attack that you prefer and also combine some of them to
perform different attacks at the same time.
.br
If a mitm method requires some parameters you can specify them after the colon.
(e.g.  \-M dhcp:ip_pool,netmask,etc )
.Sp
The following mitm attacks are available:
.RS
.TP
\fBarp\fR ([remote],[oneway])
This method implements the ARP poisoning mitm attack. ARP
requests/replies are sent to the victims to poison their ARP cache. Once the cache
has been poisoned the victims will send all packets to the attacker which, in
turn, can modify and forward them to the real destination.
.Sp
In silent mode (\-z option) only the first target is selected, if you want to
poison multiple target in silent mode use the \-j option to load a list from a
file.
.Sp
You can select empty targets and they will be expanded as 'ANY' (all the hosts in
the LAN). The target list is joined with the hosts list (created by the arp
scan) and the result is used to determine the victims of the attack.
.Sp
The parameter "remote" is optional and you have to specify it if you want to
sniff remote ip address poisoning a gateway. Indeed if you specify a victim and
the gw in the TARGETS, ettercap will sniff only connection between them, but to
enable ettercap to sniff connections that pass thru the gw, you have to use
this parameter.
.Sp
The parameter "oneway" will force ettercap to poison only from TARGET1 to
TARGET2. Useful if you want to poison only the client and not the router (where
an arp watcher can be in place).
.Sp
Example:
.Sp
the targets are: /10.0.0.1\-5/ /10.0.0.15\-20/
.br
and the host list is: 10.0.0.1 10.0.0.3 10.0.0.16 10.0.0.18
.Sp
the associations between the victims will be:
.br
1 and 16, 1 and 18, 3 and 16, 3 and 18
.Sp
if the targets overlap each other, the association with identical ip address
will be skipped.
.Sp
NOTE: if you manage to poison a client, you have to set correct routing
table in the kernel specifying the GW. If your routing table is incorrect, the
poisoned clients will not be able to navigate the Internet.


.TP
\fBicmp\fR (MAC/IP)
This attack implements ICMP redirection. It sends a spoofed icmp redirect
message to the hosts in the lan pretending to be a better route for internet. All
connections to internet will be redirected to the attacker which, in turn,
will forward them to the real gateway. The resulting attack is a HALF-DUPLEX
mitm. Only the client is redirected, since the gateway will not accept redirect
messages for a directly connected network. BE SURE TO NOT USE FILTERS THAT
MODIFY THE PAYLOAD LENGTH. you can use a filter to modify packets, but the length
must be the same since the tcp sequences cannot be updated in both ways.
.br
You have to pass as argument the MAC and the IP address of the real gateway for the lan.
.br
Obviously you have to be able to sniff all the traffic. If you are on a switch
you have to use a different mitm attack such as arp poisoning.
.Sp
NOTE: to restrict the redirection to a given target, specify it as a TARGET
.Sp
Example:
.Sp
\-M icmp:00:11:22:33:44:55/10.0.0.1
.Sp
will redirect all the connections that pass thru that gateway.


.TP
\fBdhcp\fR (ip_pool/netmask/dns)
This attack implements DHCP spoofing. It pretends to be a DHCP server and tries
to win the race condition with the real one to force the client to accept
the attacker's reply. This way ettercap is able to manipulate the GW parameter and
hijack all the outgoing traffic generated by the clients.
.br
The resulting attack is a HALF-DUPLEX mitm. So be sure to use appropriate
filters (see above in the ICMP section).
.Sp
You have to pass the ip pool to be used, the netmask and the ip of the dns server.
Since ettercap tries to win the race with the real
server, it DOES NOT CHECK if the ip is already assigned. You have to specify an
ip pool of FREE addresses to be used. The ip pool has the same form of the
target specification.
.Sp
If the client sends a dhcp request (suggesting an ip address) ettercap will ack
on that ip and modify only the gw option. If the client makes a dhcp discovery,
ettercap will use the first unused ip address of the list you have specified on
command line. Every discovery consumes an ip address. When the list is over,
ettercap stops offering new ip addresses and will reply only to dhcp requests.
.br
If you don't want to offer any ip address, but only change the router
information of dhcp request/ack, you can specify an empty ip_pool.
.Sp
BIG WARNING: if you specify a list of ip that are in use, you will mess your
network! In general, use this attack carefully. It can really mess things up!
When you stop the attack, all the victims will be still convinced that ettercap
is the gateway until the lease expires...
.Sp
Example:
.Sp
\-M dhcp:192.168.0.30,35,50\-60/255.255.255.0/192.168.0.1
.br
reply to DHCP offer and request.
.Sp
\-M dhcp:/255.255.255.0/192.168.0.1
.br
reply only to DHCP request.

.TP
\fBport\fR ([remote],[tree])
This attack implements Port Stealing. This technique is useful to sniff in a
switched environment when ARP poisoning is not effective (for example where
static mapped ARPs are used).
.Sp
It floods the LAN (based on port_steal_delay option in etter.conf) with
ARP packets. If you don't specify the "tree" option, the destination MAC
address of each "stealing" packet is the same as the attacker's one (other NICs
won't see these packets), the source MAC address will be one of the MACs in the
host list. This process "steals" the switch port of each victim host in the
host list.
Using low delays, packets destined to "stolen" MAC addresses will be received
by the attacker, winning the race condition with the real port owner.
When the attacker receives packets for "stolen" hosts, it stops the flooding
process and performs an ARP request for the real destination of the packet.
When it receives the ARP reply it's sure that the victim has "taken back" his
port, so ettercap can re-send the packet to the destination as is.
Now we can re-start the flooding process waiting for new packets.
.Sp
If you use the "tree" option, the destination MAC address of each stealing
packet will be a bogus one, so these packets will be propagated to other
switches (not only the directly connected one). This way you will be able
to steal ports on other switches in the tree (if any), but you will
generate a huge amount of traffic (according to port_steal_delay).
The "remote" option has the same meaning as in "arp" mitm method.
.Sp
When you stop the attack, ettercap will send an ARP request to each stolen
host giving back their switch ports.
.br
You can perform either HALF or FULL DUPLEX mitm according to target selection.
.Sp
NOTE: Use this mitm method only on ethernet switches. Use it carefully,
it could produce performances loss or general havoc.
.Sp
NOTE: You can NOT use this method in only-mitm mode (\-o flag), because it
hooks the sniffing engine, and you can't use interactive data injection.
.Sp
NOTE: It could be dangerous to use it in conjunction with other mitm methods.
.Sp
NOTE: This mitm method doesn't work on Solaris and Windows because of
the lipcap and libnet design and the lack of certain ioctl().
(We will feature this method on these OSes if someone will request it...)
.Sp
Example:
.Sp
The targets are: /10.0.0.1/ /10.0.0.15/
.br
You will intercept and visualize traffic between 10.0.0.1 and 10.0.0.15,
but you will receive all the traffic for 10.0.0.1 and 10.0.0.15 too.
.Sp
The target is: /10.0.0.1/
.br
You will intercept and visualize all the traffic for 10.0.0.1.

.Sp


.TP
\fBndp\fR ([remote],[oneway])
.B NOTE:
This MITM method is only supported if IPv6 support has been enabled.
.Sp
This method implements the NDP poisoning attack which is used for MITM of 
IPv6 connections. ND requests/replies are sent to the victims to poison 
their neighbor cache. Once the cache has been poisoned the victims will send 
all IPv6 packets to the attacker which, in turn, can modify and forward them 
to the real destination.
.Sp
In silent mode (\-z option) only the first target is selected, if you want to
poison multiple target in silent mode use the \-j option to load a list from a
file.
.Sp
You can select empty targets and they will be expanded as 'ANY' (all the hosts in
the LAN). The target list is joined with the hosts list (created by the arp
scan) and the result is used to determine the victims of the attack.
.Sp
The parameter "remote" is optional and you have to specify it if you want to
sniff remote ip address poisoning a gateway. Indeed if you specify a victim and
the gw in the TARGETS, ettercap will sniff only connection between them, but to
enable ettercap to sniff connections that pass thru the gw, you have to use
this parameter.
.Sp
The parameter "oneway" will force ettercap to poison only from TARGET1 to
TARGET2. Useful if you want to poison only the client and not the router (where
an arp watcher can be in place).
.Sp
Example:
.Sp
Targets are: //fe80::260d:afff:fe6e:f378/ //2001:db8::2:1/
.br
Ranges of IPv6 addresses are not yet supported.
.Sp
NOTE: if you manage to poison a client, you have to set correct routing
table in the kernel specifying the GW. If your routing table is incorrect, the
poisoned clients will not be able to navigate the Internet.
.Sp
NOTE: in IPv6 usually the link-local address of the router is being used as
the gateway address. Therefore you need to set the link-local address of the
router as one target and the global-unicast address of the victim as the other
in order to set up a successful IPv6 MITM attack using NDP poisoning.
.RE

.TP
\fB\-o\fR, \fB\-\-only\-mitm\fR

CHANGELOG

Legend:
         + new feature
         - old feature removed
         !! bug fixed
=========================================
0.8.4-nnnnnnnn    YYYYMMDD
   !! Fix fingerprint composition
   !! Default connection buffer size increased from 10KByte to 10MByte
   !! /etc/ettercap is now by default /usr/local/etc/ettercap under MacOSX, FreeBSD and Windows
   ++ new random etterfilter function
   ++ superseded legacy GeoIP C API with libmaxminddb C API

0.8.3.1-Bertillon    20200801
   !! Fix SSL protocol failure with older TLS client/server versions (min. TLS1.0)
   !! Fix blackholing SSL packets when specific redirection is used
   !! Fix TLS 1.3 interception issues (replace fake certificate with proper key length)
   !! Fix segmentation fault when parsing HTTP NTLM handshake (fixes #922)
   !! Fix crash if one redirect command is not enabled
   !! Fix build on MacOSX detecting new dependency HarfBuzz
   !! Fix warnings when parsing etter.(m)dns file when built w/o IPv6 support
   !! Fix capture delay with libpcap v1.9.1 (fixes #974)
   !! Fix segmentation fault when etterlog concatinate files
   !! Fix compiling with GCC version / defaulting to -fno-common
   !! Fix bad UDP length for packets changed with replace()
   !! Fix passing --lua-args arguments to LUA scripts
   !! Fix MSVC build when macro ORDER_ADD_{SHORT,LONG} is being used
   !! Fix references to old sourceforce.org website in the code and documentation
   !! Fix fingerprint_submit (still missing its server counterpart)
    + Take over client-side SNI extension in ClientHello in SSL interception (req. OpenSSL 1.1.1)
    + Take over SAN certificate extension from server certificate in SSL interception
    + Use server certificate sign algorithm to sign fake certificate defaulting to SHA256
    + CLI provided plugins are now also autostarted in graphical UI
    + Added --plugin-list CLI parameter
    + New execreplace etterfilter command
    + Update bundled OUI mac addresses
    + Update LuaJIT from 2.0.4 to 2.0.5
    + Update libnet from 1.1.6 to 1.2
    + Update check from 0.10.0 to 0.15.0
    + Update curl from 7.44.0 to 7.71.1
    - Separate etter.dns and etter.mdns examples in dedicated examples file
    - Remove source IP specification from customizable SSL redirects
    - Remove of deprecated redirect commands from etter.conf
    - Remove Easter Egg (Sorry ALoR and NaGA)

0.8.3-Bertillon    20190701
   !! Fix binary comparison and assignment in etterfilter
   !! Fixed packetbuffer racecond. in BRIDGE mode (e.g. Message too long)
   !! Non-aligned filters are no longer supported (recompilation with etterfilter required)
   !! Fixed sslstrip plugin startup issue due to regex compilation error
   !! Fixed lots of build warnings
   !! Proper separation of library and executable code
   !! Fixed heap-buffer-overflow in write_output in etterfilter
   !! ip_addr sanity check when etterlog processes info logfile
   !! CVE-2017-8366 (Lots of buffer under-/overflow conditions fixed)
   !! CVE-2017-6430 (Fix invalid read on crafted file in etterfilter)
   !! fix dns_spoof plugin when used in bridge mode
    + SSL redirects are now customizable at runtime
    + GeoIP detection / support using CMake
    + Rework of GTK3 UI - modern GNOME3 look
    + New Kerberos 5 downgrade plugin
    + GTK3 is the new default GTK_BUILD_TYPE
    + OSPF dissector supports more authentication methods in hash-cracker friendly format
    + Rework of Oracle O5LOGON dissector
    + Multi-threaded name resolution
    + Updated etter.finger.mac
    - GTK2 phase out initialized
    - Usage of deprecated inet_aton replaced with current successor functions

0.8.2-Ferri   20150314
   !! Fixed some openssl deprecated functions usage
   !! Fixed log file ownership
   !! Fixed mixed output print
   !! Fixed drop_privs function usage
   !! Fixed nopromisc option usage
   !! Fixed missing break in parser code
   !! Improved redirect commands
   !! Fix truncated VLAN packet headers
   !! Fix ettercap.rc file (windows only)
   !! Various cmake fixes
   !! A ton of BSD bug fixes
   !! Simplify macosx cmake files
   !! Fix incorrect sequence number after TCP injection
   !! Fix pcap length, and alignment problems with libpcap
   !! Bug fixes and gtk code refactor (gtk box wrapper)
   !! Fix some ipv6 send issues
   !! Fixed sleep time on Windows (high CPU usage)
   !! Fixed many CVE vulnerabilities (some of them already fixed in 0.8.1)
    - CVE-2014-6395 (Length Parameter Inconsistency)
    - CVE-2014-6396 (Arbitrary write)
    - CVE-2014-9376 (Negative index/underflow)
    - CVE-2014-9377 (Heap overflow)
    - CVE-2014-9378 (Unchecked return value)
    - CVE-2014-9379 (Incorrect cast)
    - CVE-2014-9380 (Buffer over-read)
    - CVE-2014-9381 (Signedness error)
    + Updated etter.finger.mac
    + Add TXT and ANY query support on dns_spoof
    + New macosx travis-ci build!
    + Enable again PDF generation
    - Remove gprof support

0.8.1-Lombroso 20141016
   !! Fixed incorrect checksum computation on 64-bit systems
   !! Fixed DNS resolution problems
   !! Fixed hurd build failure (not specific to hurd but hurd seems the first OS defining ESUCCESS in glibc)
   !! Fixed rpath handling
   !! Fixed scan host crash with recent kernels
   !! Fixed etter{log,filter} library path
   !! O5LOGON dissector fixes for stealth mode scans
   !! Fix constants to allow full hexadecimal characterset. Useful for filtering on ESP SPIs
   !! Fixed some incoherencies in gbls pointers in utils and core
   !! Fixed dhcp spoofing automatically start in text ui
   !! Many fixes in filter compiler
   !! Fixed lua installation path
   !! Many ipv6 fixes and improvements
   !! Fixed tests build failures
   !! Fixed many iconv detection problems
   !! Fixed many ctime problems
   !! Fixed many dissector ports
   !! Fixed timers incoherences
   !! Fixed powerpc build failure
   !! Fixed uniqueness of our include guards
   !! Fixed cmake warnings, by correctly linking our libraries

src/lua/share/core/ettercap.lua
---
-- Basic ettercap functionality!
--
--    Copyright (C) Ryan Linn and Mike Ryan
--
--    This program is free software; you can redistribute it and/or modify
--    it under the terms of the GNU General Public License as published by
--    the Free Software Foundation; either version 2 of the License, or
--    (at your option) any later version.
--
--    This program is distributed in the hope that it will be useful,
--    but WITHOUT ANY WARRANTY; without even the implied warranty of
--    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
--    GNU General Public License for more details.
--

--    You should have received a copy of the GNU General Public License along
--    with this program; if not, write to the Free Software Foundation, Inc.,
--    51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.

-- 


-- All of our core stuff will reside in the "ettercap" namespace.
ettercap = {}

local ffi = require("ettercap_ffi")
require('packet_meta')
local ettercap_c = require("ettercap_c")
local eclib = require("eclib")
local op = require('io')
ettercap.reg = require("ettercap_reg")

--- Log's a message using ettercap's ui_msg function.
-- @see string.format
-- @param fmt The format string
-- @param ... Variable arguments to pass in
log = function(fmt, ...) 
  -- We don't want any format string "accidents" on the C side of things.. 
  -- so, we will let lua handle it.
  ettercap_c.log(string.format(fmt, ...))
end

file_exists = function(filename)
  local msg = nil
  local fio, errmsg, errno = io.open(filename, "r")
  if fio then
    fio:close()
    return true
  end

  assert(errno == 2, "Could not open '" .. filename .. "': " .. errmsg)

  return false
end

require('dumper')

--- Dumps data structure(s) to log
-- @param ... The data to dump!
dump = function (...)
  log(DataDumper(...), "\n---")
end

-- Script interface
--
-- All Ettercap LUA scripts are initialized using a common interface. We've 
-- modeled this interface very closely after that of NMAP's NSE script 
-- interface. Our hope is that the community's familiarity with NSE will 
-- lower the barrier for entry for those looking to write Ettercap LUA 
-- scripts.
--
--
--  Data structures:
--    packet_object - Access to the Ettercap "packet_object" (originally 
--                    defined in include/ec_packet.h) is provided via a 
--                    light luajit FFI wrapper. Details on interacting with 
--                    data-types via luajit FFI can be found here:
--                    http://luajit.org/ext_ffi_semantics.html. 
--
--                    Generally, script implementations should avoid direct
--                    modifications to packet_object, or any FFI wrapped 
--                    structure, instead favoring modification through 
--                    defined ettercap.* interfaces.
--
--                    NOTE: Careful consideration must be taken to when 
--                    interacting with FFI-wrapped data-structures! Data 
--                    originating from outside of the LUA VM must have their
--                    memory managed *manually*! See the section of luajit's
--                    FFI Semantics on "Garbage Collection of cdata Objects"
--                    for details. 
--                    
--
--  Script requirements:
--
--    description - (string) Like that of NSE, each script must have a 
--                  description of the its functionality.
--
--    action      - (function (packet_object)) The action of a script operates
--                  on a FFI-wrapped packet_object. 
--
--  Optional:
--
--    packetrule  - (function (packet_object)) If implemented, then this 
--                  function must return true for a given packet_object 
--                  before that packet_object is passed to the script's action.
--

local Script = {}

do
  local coroutine = require "coroutine";
  local debug = require "debug";
  local traceback = debug.traceback;

  local ETTERCAP_SCRIPT_RULES = {
    packetrule = "packetrule",
  };

  -- These are the components of a script that are required. 
  local REQUIRED_FIELDS = {
    description = "string",
    action = "function",
--    categories = "table",
--    dependencies = "table",
  };

  function Script.new (filename, arguments)
    local script_params = arguments or {};  
    local script_path = filename 
    local full_path = ETTERCAP_LUA_SCRIPT_PATH .. "/" .. filename;

    local file_closure = nil

    if file_exists(filename) == true then
      file_closure = assert(loadfile(filename))
      script_path = filename
    elseif file_exists(full_path) == true then
      file_closure = assert(loadfile(full_path))
      script_path = full_path
    else
      log("ERROR: Could not find script '%s'\n", filename)
      return nil
    end

    local env = {
      SCRIPT_PATH = script_path,
      SCRIPT_ARGS = script_params,
      dependencies = {},
    };

    -- No idea what this does.
    setmetatable(env, {__index = _G});
    setfenv(file_closure, env);

    local co = coroutine.create(file_closure); -- Create a garbage thread
    local status, e = coroutine.resume(co); -- Get the globals it loads in env

    if not status then
      log("Failed to load %s:\n%s", filename, traceback(co, e));
      --error("could not load script");
      return nil
    end

    for required_field_name in pairs(REQUIRED_FIELDS) do
      local required_type = REQUIRED_FIELDS[required_field_name];
      local raw_field = rawget(env, required_field_name)
      local actual_type = type(raw_field);
      assert(actual_type == required_type, 
             "Incorrect of missing field: '" .. required_field_name .. "'." ..
             " Must be of type: '" .. required_type .. "'" ..
             " got type: '" .. actual_type .. "'." ..
             " Script: '" .. env["SCRIPT_PATH"] .. "'"

      );
    end

    -- Check our rules....
    local rules = {};
    for rule in pairs(ETTERCAP_SCRIPT_RULES) do
      local rulef = rawget(env, rule);
      assert(type(rulef) == "function" or rulef == nil,
          rule.." must be a function!");
      rules[rule] = rulef;
    end
    local action = env["action"];

    -- Make sure we have a hook_point!
    local hook_point = rawget(env, "hook_point")
    assert(type(hook_point) == "number", "hook_point must be a number!")
 
    local script = {
      filename = filename,
      action = action,
      rules = rules,
      hook_point = hook_point,
      env = env,
      file_closure = file_closure,
      script_params = script_params
    };
    
    return setmetatable(script, {__index = Script, __metatable = Script});
  end
end

-- Stores hook mappings.
--- Called during ettercap's shutdown
local ettercap_cleanup = function() 
end

local packet_object_ctype = ffi.typeof("struct packet_object *")
local ffi_cast = ffi.cast

local create_hook = function(script)
  local packetrule = script.rules["packetrule"]
  local hook_func = function(packet_object_ptr) 
    local packet_object = ffi_cast(packet_object_ctype, packet_object_ptr);
    if (not(packetrule == nil)) then
      if not(packetrule(packet_object) == true) then
        return false
      end
    end
    script.action(packet_object)
  end
  return(hook_func)
end

-- Adds a hook
local hook_add = function (hook_point, func)
  ettercap_c.hook_add(hook_point, func)
end

-- Processes all the --lua-script arguments into a single list of script
-- names. 
--
-- @param scripts An array of script cli arguments
-- @return A table containing all the split arguments
local cli_split_scripts = function (scripts)
  -- This keeps track of what script names have already been encountered. 
  -- This prevents us from loading the same script more than once.
  local loaded_scripts = {}

  local ret = {}
  for v = 1, #scripts do
    local s = scripts[v]
    local script_list = eclib.split(s, ",")
    for i = 1, #script_list do 
      if (loaded_scripts[script_list[i]] == nil) then
        -- We haven't loaded this script, yet, so add it it our list.
        table.insert(ret, script_list[i])
        loaded_scripts[script_list[i]] = 1
      end
    end 
  end

  return ret
end

-- Processes all the --lua-args arguments into a single list of args
-- names. 
--
-- @param args An array of args cli arguments
-- @return A table containing all the split arguments
local cli_split_args = function (args)
  local ret = {}
  for v = 1, #args do
    local s = args[v]
    local arglist = eclib.split(s, ",")
    for i = 1, #arglist do 
      -- We haven't loaded this args, yet, so add it it our list.
      local temp = eclib.split(arglist[i],"=")
      ret[temp[1]] = temp[2]
    end 
  end

  return ret
end

-- Loads a script.
--
-- @param name (string) The name of the script we want to load.
-- @param args (table) A table of key,value tuples
local ettercap_load_script = function (name, args)
  local script = assert(Script.new(name, args), "Failed to load: " .. name)
  hook_add(script.hook_point, create_hook(script))
end

-- Primary entry point for ettercap lua environment
-- @param lua_scripts Array of CLI script strings
-- @param lua_args Array of CLI argument strings
local ettercap_main = function (lua_scripts, lua_args)
  local scripts = cli_split_scripts(lua_scripts)
  local args = cli_split_args(lua_args)
  for i = 1, #scripts do
    ettercap_load_script(scripts[i], args)
  end
end

-- C -> LUA api functions. These should never be called from scripts!
ettercap.main = ettercap_main
ettercap.cleanup = ettercap_cleanup

-- Global functions

ettercap.log = log
ettercap.dump = dump

-- Is this even nescessary? Nobody should be requiring this except for 
-- init.lua... However, I'll act like this is required.
return ettercap
